core/arch/arm/include/mm/mobj.h

```diff
@@ -77,6 +78,15 @@ static inline void mobj_free(struct mobj *mobj)
 		mobj->ops->free(mobj);
 }
 
+static inline void mobj_free_wipe(struct mobj *mobj)
+{
+	void *buf = mobj_get_va(mobj, 0);
+
+	if (buf)
+		memset(buf, 0, mobj->size);
```

Comment:
wipe with 0x00 or 0x55?

Reply:
`0x00` because there is no real reason for using anything else... Yes I know, `free_wipe()` uses `0x55`, but only because this value was already used in the BGet code (when `FreeWipe` is defined).

Reply:
ok

Reply:
should this `memset()` be replaced with later introduced `memzero_explicit()`?

Reply:
Well, it seems unlikely to me that the compiler would optimize out the `memset()`, but I agree this is the kind of use case that normally calls for `memzero_explicit()`. I can introduce that function sooner in the series and use it here, no problem.

Reply:
As far as I understand `memzero_explicit()` is only needed to clean local variables/buffers on the stack. This buffer isn't stored in the stack so ordinary `memset()` should be used.

Reply:
@jenswi-linaro actually, no. The compiler can perform dead store elimination on all kinds of data, especially with Clang and LTO. Please see https://www.usenix.org/system/files/conference/usenixsecurity17/sec17-yang.pdf section 5, there are many examples.

Reply:
So now we're going to have a large grey zone where we don't know if `memset()` or `memzero_explicit()` should be used and we'll use `memzero_explicit()` just in case?

I don't see how the compiler possibly could conclude that skipping the `memset()` wouldn't be noticed this particular case.

We need guidelines for when `memzero_explicit()` should be used and when it shouldn't be used or this will become a mess. I don't mind using `memzero_explicit()` even where it's not strictly needed as long as it follows some logic.

Reply:
> We need guidelines for when `memzero_explicit()` should be used and when it shouldn't be used or this will become a mess.

+1 for guidelines. Not sure though how easy it is to specify this strictly. But IMHO guidelines should say that you should use the explicit variant if the memory being dealt with has been used for sensitive data like cryptographic keys, session objects etc.

Reply:
@jenswi-linaro @jbech-linaro the criteria I would propose for using memzero_explicit is when you are clearing/cleaning (as opposed to initializing) "sensitive" data. So that's two things really, clearing and sensitive.

Reply:
Fair enough. How about a clarification stating this in the description of `memzero_explicit()`? I think it's important to know when it's wrong to use the function also.

Reply:
Agreed, I will expand the comment.